File: src/main/java/lechuck/intellij/TaskCommandLineState.java
package lechuck.intellij;

import com.intellij.execution.ExecutionException;
import com.intellij.execution.configurations.CommandLineState;
import com.intellij.execution.configurations.GeneralCommandLine;
import com.intellij.execution.process.ProcessHandler;
import com.intellij.execution.process.ProcessHandlerFactory;
import com.intellij.execution.process.ProcessTerminatedListener;
import com.intellij.execution.runners.ExecutionEnvironment;
import lechuck.intellij.util.RegexUtil;
import org.jetbrains.annotations.NotNull;

import java.util.ArrayList;
import java.util.HashMap;

class TaskCommandLineState extends CommandLineState {
    private final TaskRunConfiguration cfg;

    public TaskCommandLineState(@NotNull ExecutionEnvironment env, TaskRunConfiguration cfg) {
        super(env);
        this.cfg = cfg;
    }

    @Override
    @NotNull
    protected ProcessHandler startProcess() throws ExecutionException {
        var processHandler = ProcessHandlerFactory.getInstance()
                .createColoredProcessHandler(createGeneralCommandLine());

        ProcessTerminatedListener.attach(processHandler);

        return processHandler;
    }

    private GeneralCommandLine createGeneralCommandLine() {
        var options = cfg.getOptions();

        var cmd = new ArrayList<String>();
        cmd.add("task");

        // taskfile
        var taskfile = options.getTaskfile();
        if (!taskfile.isEmpty()) {
            cmd.add("--taskfile");
            cmd.add(taskfile);
        }

        // task
        cmd.add(options.getTask());

        // arguments
        var arguments = options.getArguments();
        var argumentList = RegexUtil.splitBySpacePreservingQuotes(arguments);
        if (!argumentList.isEmpty()) {
            cmd.add("--");
            cmd.addAll(argumentList);
        }

        // environment variables
        var envMap = new HashMap<String, String>();

        return new GeneralCommandLine(cmd).withEnvironment(envMap);
    }
}

File: src/main/kotlin/lechuck/intellij/TaskCommandLineState.kt
package lechuck.intellij

import com.intellij.execution.configurations.CommandLineState
import com.intellij.execution.configurations.GeneralCommandLine
import com.intellij.execution.process.ProcessHandler
import com.intellij.execution.process.ProcessHandlerFactory
import com.intellij.execution.process.ProcessTerminatedListener
import com.intellij.execution.runners.ExecutionEnvironment

class TaskCommandLineState(
    env: ExecutionEnvironment,
    private val cfg: TaskRunConfiguration
) : CommandLineState(env) {

    override fun startProcess(): ProcessHandler {
        val processHandler = ProcessHandlerFactory.getInstance()
            .createColoredProcessHandler(createGeneralCommandLine())

        ProcessTerminatedListener.attach(processHandler)

        return processHandler
    }

    private fun createGeneralCommandLine(): GeneralCommandLine {
        val options = cfg.options

        val cmd = mutableListOf("task")

        // taskfile
        val taskfile = options.getTaskfile()
        if (taskfile.isNotEmpty()) {
            cmd.add("--taskfile")
            cmd.add(taskfile)
        }

        // task
        cmd.add(options.getTask())

        // environment variables
        val envMap = mutableMapOf<String, String>()

        return GeneralCommandLine(cmd)
            .withEnvironment(envMap)
    }
}